Introduce C parser library

akyrtzi opened the pull request:
Add a shared library with a C API that provides access to the syntactic parser with callbacks for the inference of raw syntax nodes.
This is primarily intended to be used by SwiftSyntax to speed-up source code parsing for it.

akyrtzi committed:
Introduce C parser library

Add a shared library with a C API that provides access to the syntactic parser with callbacks for the inference of raw syntax nodes.
This is primarily intended to be used by SwiftSyntax to speed-up source code parsing for it.

akyrtzi committed:
[utils/incrparse] Introduce testWithParserLib() helper function

This is to allow swift-syntax to run its lit tests without needing to execute swift-syntax-test
for incremental reparsing.

akyrtzi commented:
@swift-ci Please test

akyrtzi committed:
[utils/incrparse] Fix python-lint issues

akyrtzi commented:
@swift-ci test OS X platform

akyrtzi committed:
[SyntaxParser] Remove leftover type declaration and adjust the *_EXPORTS macro name

Also fixup some doc-comments.

harlanhaskins commented on tools/libSwiftSyntaxParser/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+set(LLVM_EXPORTED_SYMBOL_FILE ${CMAKE_CURRENT_SOURCE_DIR}/libSwiftSyntaxParser.exports)
+
+add_swift_host_library(libSwiftSyntaxParser SHARED
```
Nit: Formatting in this file

akyrtzi replied:
Fixed.

akyrtzi committed:
[CMake] Fix 80 col formatting

akyrtzi committed:
[CMake] Add 'parser-lib' install component and install the relevant files

The files go into 'lib/swift' so that SwiftSyntax can import the library without needing custom search paths to be added to the swiftpm invocation.

nkcsgexi commented on tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp:
```diff
@@ -0,0 +1,216 @@
+  SourceManager &SM;
+  unsigned BufferID;
+  swiftparse_node_handler_t NodeHandler;
+  swiftparse_node_lookup_t NodeLookup;
```
Could we have include them with a `SynParser` here?

akyrtzi replied:
Done.

nkcsgexi commented on tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp:
```diff
@@ -0,0 +1,216 @@
+public:
+  explicit CLibParseActions(SourceManager &sm, unsigned bufID,
+                            swiftparse_node_handler_t hdl,
+                            swiftparse_node_lookup_t lookup)
```
Similarly, this constructor can take a `SynParser&` as the third parameter.

akyrtzi replied:
Done.

akyrtzi committed:
[libSwiftSyntaxParser] Pass SynParser to CLibParseActions to use as container for the node handling blocks

akyrtzi committed:
[SwiftSyntaxParser] Add doc-comments for swiftparse_parser_t

akyrtzi committed:
[SwiftSyntaxParser] Clarify doc-comment about the order of the nodes that the parser provides

akyrtzi committed:
[utils/build-script-impl] Pass header/lib paths to the syntax parser library for the build-script invocation of SwiftSyntax

akyrtzi commented:
@swift-ci Please test

akyrtzi commented:
@swift-ci test OS X platform
